src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.cpp

```diff
@@ -0,0 +1,487 @@
+ */
+void FlybyODuKF::UpdateState(uint64_t CurrentSimNanos)
+{
+    double newTimeTag;  /* [s] Local Time-tag variable*/
+    bool computePostFits = false;
+
+    /*! - If the time tag from the measured data is new compared to previous step,
+     propagate and update the filter*/
+    this->opNavHeadingBuffer = this->opNavHeadingMsg();
+    newTimeTag = this->opNavHeadingBuffer.timeTag * NANO2SEC;
+    if(newTimeTag >= this->timeTag && this->opNavHeadingBuffer.valid)
+    {
+        TimeUpdate(newTimeTag);
+        MeasurementUpdate();
+        computePostFits = true;
+    }
+    /*! - If current clock time is further ahead than the measured time, then
+     propagate to this current time-step*/
+    newTimeTag = CurrentSimNanos*NANO2SEC;
+    if(newTimeTag >= this->timeTag)
+    {
+        TimeUpdate(newTimeTag);
+    }
+
+    this->opNavFilterMsgBuffer = this->opNavFilterMsg.zeroMsgPayload;
+    this->navTransOutMsgBuffer = this->navTransOutMsg.zeroMsgPayload;
+
+    /*! - The post fits are y - ybar if a measurement was read, if observations are zero, do not compute post fit residuals*/
+    if(computePostFits){
+        /*! - Compute Post Fit Residuals, first get Y (eq 22) using the states post fit*/
+        MeasurementModel();
+        /*! - Compute the value for the yBar parameter (equation 23)*/
+        this->postFits.setZero(this->obs.size());
+        Eigen::VectorXd yBar;
+        yBar.setZero(this->obs.size());
+        for(int i=0; i<this->numberSigmaPoints; i++){
+            yBar += this->wM(i)*this->yMeas.col(i);
+        }
+        this->postFits = this->obs - yBar;
+        eigenMatrixXd2CArray(this->postFits, this->opNavFilterMsgBuffer.postFitRes);
+    }
+
+    /*! - Write the flyby OD estimate into the copy of the navigation message structure*/
+    eigenMatrixXd2CArray(1e3*this->state.head(3), this->navTransOutMsgBuffer.r_BN_N);
+    eigenMatrixXd2CArray(1e3*this->state.tail(3), this->navTransOutMsgBuffer.v_BN_N);
+
+    /*! - Populate the filter states output buffer and write the output message*/
+    this->opNavFilterMsgBuffer.timeTag = this->timeTag;
+    eigenMatrixXd2CArray( 1e3*this->state, this->opNavFilterMsgBuffer.state);
+    eigenMatrixXd2CArray( 1e3*this->xBar, this->opNavFilterMsgBuffer.stateError);
+    eigenMatrixXd2CArray(1e6*this->covar, this->opNavFilterMsgBuffer.covar);
+
+    this->navTransOutMsg.write(&this->navTransOutMsgBuffer, this->moduleID, CurrentSimNanos);
+    this->opNavFilterMsg.write(&this->opNavFilterMsgBuffer, this->moduleID, CurrentSimNanos);
+}
+
+/*! This method performs the time update for the flyby OD kalman filter.
+ It propagates the sigma points forward in time and then gets the current
+ covariance and state estimates.
+ @return void
+ @param configData The configuration data associated with the OD filter
+ @param updateTime The time that we need to fix the filter to (seconds)
+ */
+void FlybyODuKF::TimeUpdate(double updateTime)
+{
+    size_t i;
```

Comment:
:large_orange_diamond: inline loop counters. This is a good safe practice as it limits the scope of the variable and allows the compiler to take advantage of potential optimizations.

Reply:
Done